### PR TITLE
Add RPC bench threads to accounts-cluster-bench

### PR DESCRIPTION
**Pull request by sakridge**
#### Problem

No way to benchmark the RPC subsystem while creating large numbers of accounts like with accounts-cluster-bench.

#### Summary of Changes

Add RPC benchmarking threads to accounts-cluster-bench which can query for the created accounts while the bench is running.

Fixes #

### CODE REVIEW

File: accounts-cluster-bench/src/main.rs [hunk 1]
```diff
@@ -619,6 +902,15 @@ fn main() {
         eprintln!("bad num_instructions: {num_instructions}");
         exit(1);
     }
+    let num_rpc_bench_threads = value_t!(matches, "num_rpc_bench_threads", usize).unwrap_or(0);
+    let rpc_benches = values_t!(matches, "rpc_bench", String)
```
Comment (alessandrod) [inline accounts-cluster-bench/src/main.rs:906]: I don't think you've actually added these two options to clap?

Comment (CriesofCarrots) [reply accounts-cluster-bench/src/main.rs:906]: ☝️

File: accounts-cluster-bench/src/main.rs [hunk 2]
```diff
@@ -227,6 +229,264 @@ fn make_close_message(
...
+                        stats.total_success_time_us += rpc_time.as_us();
+                    }
+                    Err(e) => {
+                        rpc_time.stop();
+                        stats.total_errors_time_us += rpc_time.as_us();
+                        stats.errors += 1;
+                        if last_error.elapsed().as_secs() > 2 {
+                            info!("get_slot error: {:?}", e);
+                            last_error = Instant::now();
+                        }
```
Comment (CriesofCarrots) [inline accounts-cluster-bench/src/main.rs:357]: Maybe dedupe this into a closure or fn in case we want to twiddle the frequency or make other changes to all RpcBench variants?

File: accounts-cluster-bench/src/main.rs [hunk 3]
```diff
@@ -227,6 +229,264 @@ fn make_close_message(
...
+            }
+            RpcBench::Version => match client.get_version() {
+                Ok(_r) => {
+                    stats.success += 1;
+                }
+                Err(_e) => {
+                    stats.errors += 1;
+                }
+            },
+        }
```
Comment (CriesofCarrots) [inline accounts-cluster-bench/src/main.rs:417]: Why not count `total_success_time_us` and `total_errors_time_us` for either of these variants?

File: accounts-cluster-bench/src/main.rs [hunk 4]
```diff
@@ -227,6 +229,264 @@ fn make_close_message(
...
+                    }
+                }
+                stats.total_success_time_us += rpc_time.as_us();
+                stats.success += 1;
+            }
+            Err(e) => {
+                rpc_time.stop();
+                stats.total_errors_time_us += rpc_time.as_us();
+                stats.errors += 1;
+                if last_error.elapsed().as_secs() > 1 {
```
Comment (CriesofCarrots) [inline accounts-cluster-bench/src/main.rs:304]: Is this intended to only be 1sec? The other elapsed checks are 2sec.

File: local-cluster/tests/local_cluster.rs [hunk 5]
```diff
@@ -240,10 +240,7 @@ fn test_local_cluster_signature_subscribe() {
     );
 
     let (mut sig_subscribe_client, receiver) = PubsubClient::signature_subscribe(
-        &format!(
-            "ws://{}",
-            &non_bootstrap_info.rpc_pubsub().unwrap().to_string()
-        ),
+        &format!("ws://{}", non_bootstrap_info.rpc_pubsub().unwrap()),
```
Comment (CriesofCarrots) [inline local-cluster/tests/local_cluster.rs:243]: This fmt change seems unrelated?

File: accounts-cluster-bench/src/main.rs [hunk 6]
```diff
@@ -227,6 +229,264 @@ fn make_close_message(
...
+
+impl FromStr for RpcBench {
+    type Err = RpcParseError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "get-slot" => Ok(RpcBench::GetSlot),
+            "get-multiple-accounts" => Ok(RpcBench::GetMultipleAccounts),
+            "get-token-accounts-by-owner" => Ok(RpcBench::GetTokenAccountsByOwner),
+            "version" => Ok(RpcBench::Version),
```
Comment (CriesofCarrots) [inline accounts-cluster-bench/src/main.rs:254]: Technically, version also has a `get` prefix in RPC: `getVersion`. Wdyt about removing all the `get`s from expected strings as well as from the `RpcBench` variants?

**Review by CriesofCarrots — COMMENTED** [review]
Generally looks fine. Just a couple thoughts and questions.

File: accounts-cluster-bench/src/main.rs [hunk 7]
```diff
@@ -291,6 +553,22 @@ fn run_accounts_bench(
         None,
     );
 
+    let exit = Arc::new(AtomicBool::new(false));
+    let base_keypair_pubkey = base_keypair.pubkey();
+    let rpc_bench_threads: Vec<_> = if let Some(rpc_benches) = rpc_benches {
+        make_rpc_bench_threads(
```
Comment (alessandrod) [inline accounts-cluster-bench/src/main.rs:559]: These threads will start racing with account creation right? So stats.errors
wont' be very accurate since it'll depend on scheduling etc. Maybe move
starting the threads right after account creation?

File: accounts-cluster-bench/src/main.rs (hunk 3 above)
Comment (sakridge) [reply accounts-cluster-bench/src/main.rs:417]: just an oversight, i'll add them

File: accounts-cluster-bench/src/main.rs (hunk 6 above)
Comment (sakridge) [reply accounts-cluster-bench/src/main.rs:254]: ok, sure. I actually added Version because clippy doesn't like the Get prefix and I thought it might be useful as well, but I can just remove the 'get' prefix.

File: local-cluster/tests/local_cluster.rs (hunk 5 above)
Comment (sakridge) [reply local-cluster/tests/local_cluster.rs:243]: yea, some cleanup i'll move to another PR

File: accounts-cluster-bench/src/main.rs (hunk 4 above)
Comment (sakridge) [reply accounts-cluster-bench/src/main.rs]: changed to 2 to match the others

File: accounts-cluster-bench/src/main.rs (hunk 2 above)
Comment (sakridge) [reply accounts-cluster-bench/src/main.rs:357]: I can do this in a follow-up, requires some re-shuffling.

**Review by CriesofCarrots — APPROVED** [review]
